### ISSUE TITLE
Endless loop on startup when built with -static (MSYS2/Windows 10)

### ISSUE THREAD

**Issue by vsplesk** [ccache/ccache#1009]
[Logfile_filtered.csv](https://github.com/ccache/ccache/files/8111193/Logfile_filtered.csv)
### How to reproduce ###

1. Install latest MSYS2, with gcc and base-devel and ninja (which brings gcc 11.x)
2. Configure with -DCMAKE_BUILD_TYPE=MinSizeRel -DCMAKE_CXX_FLAGS=-static -DZSTD_FROM_INTERNET=ON -G Ninja -DREDIS_STORAGE_BACKEND=OFF
3. Build with "cmake --build ."
4. Run ./ccache.exe --help (which works)
5. Configure a real project (cmake, with CMACE_CXX_LAUNCHER option pointing to the new ccache.exe)

### Actual behavior ###
ccache gets stuck, showing 100% CPU load on every core

### Expected behavior ###
ccache shall work. It did work the last time I tested the code from PR #942 (Tricore support), however it was compiled with MSVC last time (which did work last year but now MSVC build seems to be totally broken, hence switching to MSYS2 toolchain).

Also see attachment for the monitored system activities - it seems to keep repeating the last ~25-30 operations again and again and again.

### Environment ###
Windows 10 x64, build 1809. Calling the user build process from a git-bash.
See above for other tools involved.

ccache version 4.x
(from current master, https://github.com/ccache/ccache/commit/22a87e468e721649a622ee894e692963d5d4114c )

**Comment by jrosdahl** [comment]
That system activity log unfortunately doesn't tell me anything.

@rkitover, @rvogg, @cristianadam, @llunak: Any ideas here?

**Comment by rkitover** [comment]
I will look, I don't think you're supposed to use `-static` on Windows, just link to static libs, at least I've never done it that way, I don't know.

I will look at fixing MSVC support too, I did that last time, unfortunately everything bitrots.

**Comment by vsplesk** [comment]
Regular MSVC build did not work for me for obscure reasons - failed to link, references to some blake3* code were not resolvable. Although the generated msbuild setup looked okay at the first glance.

Meanwhile I repeated the action on a totally fresh Windows installation. Got MSYS2 and mingw-w64 toolchain (howto from msys2.org) plus cmake and libzstd-devel. Result: same mess, nothing works as soon as a file is specified. Tried to observe with gdb but the issue is elusive, the interrupt often stops "somewhere" and the stack does not tell me much.

So, out of desperation and assuming that your regular builds might be done in container with cross-platform builds, I have compiled the same with mingw-64 toolchain (from Debian Testing) and see: that executable works on Windows without issues.

Remarks:
a) even with MinSizeRel build type, my binary is slightly larger than 2MB. I am wondering how you get it to that size . Is the release ZIP just the result of the workflows?
b) it would be good to mention .github/workflows/build.yaml somewhere. It DID take me a while to figure out why the build is failing with scariest symptoms. Using x86_64-w64-mingw32-g++-posix instead of x86_64-w64-mingw32-g++ is the key. I will probably add a change PR for that.

**Comment by rkitover** [comment]
@vsplesk This is the right way to make a static build using MSYS2 and your settings:

```bash
cmake .. -DCMAKE_BUILD_TYPE=MinSizeRel -DZSTD_FROM_INTERNET=ON -DREDIS_STORAGE_BACKEND=OFF -DCMAKE_FIND_LIBRARY_SUFFIXES=.a -DCMAKE_LINK_SEARCH_START_STATIC=ON -DCMAKE_LINK_SEARCH_END_STATIC=ON -DCMAKE_CXX_COMPILER_LAUNCHER=$HOME/ccache.exe -G Ninja
```

, here you can see I'm using a copy of ccache that was built this way and it works fine.

Here the output of `dumpbin /dependents` for the executable I built:

```console
Dump of file ccache.exe

File Type: EXECUTABLE IMAGE

  Image has the following dependencies:

    ADVAPI32.dll
    KERNEL32.dll
    msvcrt.dll
    WS2_32.dll
```
.

**Comment by rkitover** [comment]
MSVC build also still works fine, this is the `cmake` command I used:

```powershell
cmake .. -DCMAKE_BUILD_TYPE=MinSizeRel -DZSTD_FROM_INTERNET=ON -DREDIS_STORAGE_BACKEND=OFF -DCMAKE_FIND_LIBRARY_SUFFIXES='.lib' -DCMAKE_LINK_SEARCH_START_STATIC=ON -DCMAKE_LINK_SEARCH_END_STATIC=ON -G Ninja
```

**Comment by rkitover** [comment]
Here's the output of `dumpbin /dependents` for the MSVC-built binary:

```console
Dump of file ccache.exe

File Type: EXECUTABLE IMAGE

  Image has the following dependencies:

    WS2_32.dll
    KERNEL32.dll
    ADVAPI32.dll
    MSVCP140.dll
    MSVCP140_CODECVT_IDS.dll
    VCRUNTIME140.dll
    VCRUNTIME140_1.dll
    api-ms-win-crt-runtime-l1-1-0.dll
    api-ms-win-crt-math-l1-1-0.dll
    api-ms-win-crt-string-l1-1-0.dll
    api-ms-win-crt-time-l1-1-0.dll
    api-ms-win-crt-stdio-l1-1-0.dll
    api-ms-win-crt-environment-l1-1-0.dll
    api-ms-win-crt-filesystem-l1-1-0.dll
    api-ms-win-crt-heap-l1-1-0.dll
    api-ms-win-crt-convert-l1-1-0.dll
    api-ms-win-crt-utility-l1-1-0.dll
    api-ms-win-crt-locale-l1-1-0.dll
```

these are just runtime DLLs, there's a way to link the runtime statically too with an MSVC flag.

**Comment by jrosdahl** [comment]
@rkitover: Thanks!

@vsplesk wrote:

> a) even with MinSizeRel build type, my binary is slightly larger than 2MB. I am wondering how you get it to that size . Is the release ZIP just the result of the workflows?

The Windows binary releases are created by me by cross-compiling on my local machine as part of the release process. The CMake commands I use are these:

    CC=i686-w64-mingw32-gcc-posix CXX=i686-w64-mingw32-g++-posix \
      cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_SYSTEM_NAME=Windows \
            -DZSTD_FROM_INTERNET=ON -DHIREDIS_FROM_INTERNET=ON
    CC=x86_64-w64-mingw32-gcc-posix CXX=x86_64-w64-mingw32-g++-posix \
      cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_SYSTEM_NAME=Windows \
            -DZSTD_FROM_INTERNET=ON -DHIREDIS_FROM_INTERNET=ON

**Comment by rkitover** [comment]
@jrosdahl How do you feel about me making a PR for a `-DENABLE_SHARED=FALSE` flag, which would set the options above properly as well as handle linking the runtime statically on MSVC (which cannot be done on the commandline.)

**Comment by cristianadam** [comment]
> @jrosdahl How do you feel about me making a PR for a `-DENABLE_SHARED=FALSE` flag, which would set the options above properly as well as handle linking the runtime statically on MSVC (which cannot be done on the commandline.)

Try with `-DCMAKE_MSVC_RUNTIME_LIBRARY=MultiThreaded` and you should get a MSVC binary linked statically to the runtime library.

**Comment by rkitover** [comment]
> Try with `-DCMAKE_MSVC_RUNTIME_LIBRARY=MultiThreaded` and you should get a MSVC binary linked statically to the runtime library.

Unfortunately this did not work for me, I believe this is a property only and cannot be set with a global variable.

**Comment by cristianadam** [comment]
> > Try with `-DCMAKE_MSVC_RUNTIME_LIBRARY=MultiThreaded` and you should get a MSVC binary linked statically to the runtime library.
> 
> Unfortunately this did not work for me, I believe this is a property only and cannot be set with a global variable.

`CMAKE_MSVC_RUNTIME_LIBRARY` contains the default value that CMake uses to populate the target's `MSVC_RUNTIME_LIBRARY` property.

It's available since CMake 3.15, see https://cmake.org/cmake/help/latest/variable/CMAKE_MSVC_RUNTIME_LIBRARY.html.

Which generator are ou using? Should work with Ninja.

**Comment by rkitover** [comment]
@vsplesk I've reproduced your MSVC build issue, it seems to happen with the default `msbuild` generator, as opposed to Ninja. I will try to take a look later.

![ccache msbuild error](https://i.imgur.com/s5NTe4d.jpg)

**Comment by rkitover** [comment]
If I wrap the `IF()` statement [here](https://github.com/ccache/ccache/blob/0b3c5f17de5af6b5f0520295d96eb21c6ebe5a60/src/third_party/blake3/CMakeLists.txt#L30) in:

```cmake
if(NOT MSVC)
...
endif()
```

then the `msbuild` build works fine. I'm looking into the best way to fix this.

**Comment by rkitover** [comment]
> `CMAKE_MSVC_RUNTIME_LIBRARY` contains the default value that CMake uses to populate the target's `MSVC_RUNTIME_LIBRARY` property.
> 
> It's available since CMake 3.15, see https://cmake.org/cmake/help/latest/variable/CMAKE_MSVC_RUNTIME_LIBRARY.html.
> 
> Which generator are ou using? Should work with Ninja.

In this project it seems to not be working neither with Ninja nor msbuild.

**Comment by rkitover** [comment]
The fix for MSVC turned out to be trivial, see my [pr](https://github.com/ccache/ccache/pull/1043), just adding `ASM` and `ASM_MASM` to `project()` languages, this does not affect the Linux build which I tested as well.

**Comment by rkitover** [comment]
@jrosdahl I see now you have a `STATIC_LINK` option, any objections if I make some updates to this and also add MSVC support?

**Comment by jrosdahl** [comment]
> @jrosdahl I see now you have a STATIC_LINK option, any objections if I make some updates to this and also add MSVC support?

Sounds good to me.

**Comment by jrosdahl** [comment]
@rkitover: Can this be closed now (because of #1058) or is there more to do?

**Comment by rkitover** [comment]
@jrosdahl

I believe so.

Windows builds will now be static by default if all static libraries are available, and in the case of zstd/hireds FROM_INTERNET those are as well, so no special cmake flags or other settings are needed any more for this.

I was also thinking it would be nice to enable the two FROM_INTERNET flags automatically if system libraries are not found. Some people might not like the idea of automatically using external sources, e.g. package builders, but in those scenarios internet access is usually disabled anyway. We could also print a loud warning e.g. **\*\*\* DOWNLOADING MISSING PACKAGE &lt;PKG&gt; FROM INTERNET...** What do you think? This would allow automatic builds with no flags in most scenarios.

The strange behavior with the `-static` flag reported here is probably some compiler/linker weirdness that we can do nothing about, or some other bug somewhere.

@vsplesk any objections?

**Comment by rkitover** [comment]
@jrosdahl 

Since there is no response from the submitter, I think this issue can be closed.

What do you think about my automatic FROM_INTERNET idea?

**Comment by vsplesk** [comment]
Tried again, even without -static flag. Just compiling with default settings.
The resulting binary works again, thank you.

Also tried the MSVC build (disabling Redis there, though). The linking fixes from #1058 seem to have worked, no missing blake* symbols anymore. Thanks.

**Comment by jrosdahl** [comment]
> What do you think about my automatic FROM_INTERNET idea?

I would be OK with that.
